File: scripts/bot_fortress.lua
-- bot_fortress.lua — Tracked laser fortress.
-- Build: tracks + tank body + 4 weapons (the maximum):
--   Laser (left), Laser (right), Laser (top), AutoCannon (top_rear).
-- The heaviest chassis. Almost immobile but pivots fast on its tracks.
-- Wins by continuous beam damage from three lasers; the rear AutoCannon
-- punishes anything that gets behind it.

function init()
    return {
        locomotion = "tracks",
        body       = "tank",
        weapons = {
            { type = "Laser",      mount = "left"     },
            { type = "Laser",      mount = "right"    },
            { type = "Laser",      mount = "top"      },
            { type = "AutoCannon", mount = "top_rear" },
        },
    }
end

math.randomseed()

local wander_angle = math.random() * math.pi * 2
local wander_timer = 0.0

local function wall_avoid(targets)
    local ax, az = 0, 0
    for _, r in ipairs(targets) do
        if r.type == "wall" and r.distance < 2.5 then
            local dx = self_x - r.x
            local dz = self_z - r.z
            local d  = math.max(r.distance, 0.05)
            ax = ax + dx / (d * d)
            az = az + dz / (d * d)
        end
    end
    return ax, az
end

function think(dt)
    wander_timer = wander_timer - dt

    local targets = scan(28.0)
    local ax, az  = wall_avoid(targets)

    local enemy, min_dist = nil, math.huge
    for _, t in ipairs(targets) do
        if t.type == "bot" and t.team == 6 and t.distance < min_dist then
            min_dist = t.distance
            enemy    = t
        end
    end

    if enemy then
        local dx = enemy.x - self_x
        local dz = enemy.z - self_z
        local mx = dx + ax
        local mz = dz + az
        move(mx, mz)
        -- All three lasers stay locked on; the AutoCannon (slot 4) only
        -- fires when the target is well in range, since its 0.6s reload
        -- otherwise wastes shots.
        fire_weapon(1, dx, dz)
        fire_weapon(2, dx, dz)
        fire_weapon(3, dx, dz)
        if enemy.distance < 14.0 then
            fire_weapon(4, dx, dz)
        end
    else
        local al = math.sqrt(ax * ax + az * az)
        if al > 0.4 then
            wander_angle = math.atan(az, ax) + (math.random() - 0.5) * 0.5
            wander_timer = 0.8
        elseif wander_timer <= 0.0 then
            wander_angle = math.random() * math.pi * 2
            wander_timer = 1.8 + math.random()
        end
        local wx = math.cos(wander_angle) + ax
        local wz = math.sin(wander_angle) + az
        move(wx, wz)
    end
end

File: scripts/bot_fortress.py
# bot_fortress.py — Tracked laser fortress.
# Build: tracks + tank body + 4 weapons (maximum):
#   Laser (left), Laser (right), Laser (top), AutoCannon (top_rear).
# Heaviest chassis. Almost immobile but pivots fast on its tracks.
# Continuous beam damage from three lasers; rear AutoCannon for close threats.
# fire_weapon uses 0-based index: 0=Laser/left, 1=Laser/right, 2=Laser/top, 3=AutoCannon/top_rear

import math
import random


def init():
    return {
        "locomotion": "tracks",
        "body": "tank",
        "weapons": [
            {"type": "Laser",      "mount": "left"},
            {"type": "Laser",      "mount": "right"},
            {"type": "Laser",      "mount": "top"},
            {"type": "AutoCannon", "mount": "top_rear"},
        ],
    }


_wander_angle = random.uniform(0, math.pi * 2)
_wander_timer = 0.0


def _wall_avoid(targets):
    ax, az = 0.0, 0.0
    for r in targets:
        if r["type"] == "wall" and r["distance"] < 2.5:
            dx = self_x - r["x"]
            dz = self_z - r["z"]
            d = max(r["distance"], 0.05)
            ax += dx / (d * d)
            az += dz / (d * d)
    return ax, az


def think(dt):
    global _wander_angle, _wander_timer

    _wander_timer -= dt

    targets = scan(28.0)
    ax, az = _wall_avoid(targets)

    enemy, min_dist = None, float("inf")
    for t in targets:
        if t["type"] == "bot" and t["team"] == 6 and t["distance"] < min_dist:
            min_dist = t["distance"]
            enemy = t

    if enemy:
        dx = enemy["x"] - self_x
        dz = enemy["z"] - self_z
        move(dx + ax, dz + az)
        # Three lasers locked on at all times
        fire_weapon(0, dx, dz)
        fire_weapon(1, dx, dz)
        fire_weapon(2, dx, dz)
        # AutoCannon only fires within effective range (0.6s reload)
        if enemy["distance"] < 14.0:
            fire_weapon(3, dx, dz)
    else:
        al = math.sqrt(ax * ax + az * az)
        if al > 0.4:
            _wander_angle = math.atan2(az, ax) + (random.random() - 0.5) * 0.5
            _wander_timer = 0.8
        elif _wander_timer <= 0.0:
            _wander_angle = random.uniform(0, math.pi * 2)
            _wander_timer = 1.8 + random.random()
        wx = math.cos(_wander_angle) + ax
        wz = math.sin(_wander_angle) + az
        move(wx, wz)
